lib/report_portal/cucumber/report.rb

```diff
@@ -46,19 +46,57 @@ def initialize
       end
 
       def start_launch(desired_time = ReportPortal.now, cmd_args = ARGV)
+        # Not sure what is the use case if launch id is missing. But it does not make much of practical usage
+        #
+        # Expected behavior that make sense:
+        #  1. If launch_id present attach to existing (simple use case)
+        #  2. If launch_id not present check if exist rp_launch_id.tmp
+        #  3. [ADDED] If launch_id is not present check if lock exist with launch_uuid
         if attach_to_launch?
           ReportPortal.launch_id =
               if ReportPortal::Settings.instance.launch_id
                 ReportPortal::Settings.instance.launch_id
               else
-                file_path = ReportPortal::Settings.instance.file_with_launch_id || (Pathname(Dir.tmpdir) + 'rp_launch_id.tmp')
-                File.read(file_path)
+                file_path = lock_file
+                if File.file?(file_path)
+                  File.read(file_path)
+                else
+                  new_launch(desired_time, cmd_args, file_path)
```

Comment:
Why do consider it useful to create a launch if user told that process should write results to an existing launch?

Reply:
Why does user need to manage launches separately? 
If you want to run tests in parallel, you should be able simply to run it as before you add integration with report portal. agent_ruby is just another formatter.
That is all I added to our existing tests:
```rb
AfterConfiguration do |config|
  formats = config.formats
  formats.push(["ReportPortal::Cucumber::ParallelFormatter", {}, set_up_output_format(report_name, :report_portal)])
  # REPORT PORTAL
  ENV['rp_endpoint'] = 'http://k8devop......cloud:32561/api/v1'
  ENV['rp_disable_ssl_verification'] = 'true'
  ENV['rp_is_debug']= 'true' unless ENV['LOGNAME']
  ENV['rp_launch'] = report_name
  ENV['rp_project'] = QA_PROJECT
  report_portal_tags = ['cucumber', QA_SUB_PROJECT,ENV['TEST_ENV'].split('.')[0]]
  report_portal_tags.push(ENV['BUILD_TAG']) if ENV['BUILD_TAG']
  report_portal_tags.push(ENV['LOGNAME']) if ENV['LOGNAME']
  ENV['rp_tags'] =report_portal_tags.to_s
  ENV['TMPDIR'] = '.tmp'
  ENV['rp_formatter_modes'] ='attach_to_launch'
  env = Cukeconfig::Config.instance.get_environment
  ENV['rp_uuid'] = env[:accounts][:report_portal][:uuid]
  ENV['rp_launch_uuid'] =  ENV['RP_REPORT_UUID'] ? ENV['RP_REPORT_UUID'] : SecureRandom.uuid

  Commapny::Tools::ReportLogger.instance.configure(formats: config.formats, profile: config.to_hash[:profiles][0])
end
```

Reply:
Well, it looks like you want to create a launch when starting running features. And close it when you finish running features.
In that case you don't need to use `attach_to_launch`.
`attach_to_launch` mode is for a situation when the launch is created e.g. by rake task before running features, and closed e.g. by rake task after running features.

Reply:
with parallel_tests cucumber executed in separate processes, and it only creating new launch if rp_launch_id not provided (by the first process).

How would you implement parallel execution if you don't have a launch created beforehand?

Reply:
> How would you implement parallel execution if you don't have a launch created beforehand?

If `attach_to_launch` formatter mode isn't specified, intended / probably current behaviour looks to be the following.
The first Cucumber process creates a launch. Other Cucumber processes wait for that launch to be created.
Then tests run as usual.
When tests finish, all processes but first don't finish the launch. When the first process gets finished, it waits for all other processes to finish, and then finishes the launch.

Reply:
My previous comment is for situation when tests are executed like:
> parallel_cucumber <some options> -o '<some other options> -f ReportPortal::Cucumber::ParallelFormatter'

Reply:
> The first Cucumber process creates a launch. Other Cucumber processes wait for that launch to be created.
this is not true:
https://github.com/reportportal/agent-ruby/blob/01fbe3a8261194f55cc96029595266583875fc54/lib/report_portal/cucumber/report.rb#L50-L73

it will not create lock file and force each process to create its own launch id

Reply:
My comment referred to the code in this commit - https://github.com/reportportal/agent-ruby/blob/e1c4805d7181658da243ce1872e4218a7b88cc0a/lib/report_portal/cucumber/parallel_report.rb#L36-L58